clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```diff
@@ -989,6 +989,18 @@ static mlir::Value maybeBuildArrayDecay(mlir::OpBuilder &builder,
   assert(arrayPtrTy && "expected pointer type");
   auto arrayTy = arrayPtrTy.getPointee().dyn_cast<::mlir::cir::ArrayType>();
 
+  // FIXME(cir): the check below is rather unintuitive and may not work in all
+  // scenarios. The issue is that this method has no way to check if the
+  // subscript operator is being applied to a pointer or an array. In the first
+  // case, it wrongly decays multi-dim arrays twice before applying the first
+  // dimension's stride, generating a invalid access. Ideally, I think we should
+  // stray from the original code gen and have two separate methods for this:
+  // buildArraySubscript and buildPointerSubscript. One would decay, the other,
```

Comment:
That'd work for me, we should always deviate from original codegen when it makes sense, as long as we document so we don't lose ourselves in the future.